src/wp-includes/functions.php

```diff
@@ -7635,7 +7635,7 @@ function recurse_dirsize( $directory, $exclude = null, $max_execution_time = nul
 		$save_cache      = true;
 	}
 
-	if ( isset( $directory_cache[ $cache_path ] ) ) {
+	if ( isset( $directory_cache[ $cache_path ] ) && is_int( $directory_cache[ $cache_path ] ) ) {
 		return $directory_cache[ $cache_path ];
 	}
```

Comment:
In two minds on this suggestion but the theory is to:
- avoid running computationally heavy code below if the data is available in the older format
- remove the old format data from the transient

The latter is more important to avoid database bloat, the transient just fills and fills. This will also be an issue with the switch to `WP_CONTENT` but less  easily avoided. 

```suggestion
	if ( isset( $directory_cache[ $cache_path ] ) ) {
		if (
			isset( $directory_cache[ $cache_path ]['size'] ) &&
			is_int( $directory_cache[ $cache_path ]['size'] )
		) {
			$directory_cache[ $cache_path ] = $directory_cache[ $cache_path ]['size'];
			if ( $save_cache ) {
				set_transient( 'dirsize_cache', $directory_cache );
			}
		}

		if ( is_int( $directory_cache[ $cache_path ] ) ) {
			return $directory_cache[ $cache_path ];
		}
	}
```

Reply:
🤔 , that's a great point. i'm more concerned with the latter as well, but would love to hear what @JanThiel thinks about the performance implications.

Would it be simpler to just delete the whole transient when the old format is detected, though? Either with something in this function, or through the upgrade routine?

Reply:
I reckon remove it as part of the upgrade routine. 

We can still account for the older format here for sites that don't run the upgrade routine. Anecdotally this is larger sites so they're more likely to have the transient fall out of an object cache at some point too.

Reply:
🤔 , yeah. I wonder if the upgrade routine is even worth it, then? Doesn't seem like it if we're gonna handle it here, and there's half a chance this code will have already migrated it by the time someone runs the routine.

My gut leans towards keeping this simple, but the added code here isn't really all that complex, and I agree that it's addressing valid concerns.

Reply:
I've pushed the legacy handling including the above in ad28b2042ee62bfe0477d4f6be24ace43ef0f780 and bd749dc5e96f6a9e07b1bc50183a45995dba7fbc. 

I am in two minds as to whether we need both this and the upgrade routine. For most typical sites the logic here won't be used. I am happy either way if your night brain forms an opinion.

Reply:
Keep in mind that the current WP5.5  transient should just live 1 hour. It was never ment to really _last long_.
So I would suggest to simply remove the transient, when old data is detected. This will be a solid fallback for people not upgrading the DB, as well as custom code or plugins still creating the old transient format. Maybe add a log warning hinting that there might be some code doing old stuff. 

So what about this?

```suggestion
	if ( isset( $directory_cache[ $cache_path ] ) ) {
                if ( is_int( $directory_cache[ $cache_path ] ) ) {
		        return $directory_cache[ $cache_path ];
	        } else {
	                // Found malformed (maybe pre WP5.6?) content in dirsize cache. Reset it.
	                $directory_cache = array();
	                delete_transient( 'dirsize_cache' );
	                // Maybe log out a Warning as well?
	        }
	}
```

Reply:
> the current WP5.5 transient should just live 1 hour

That's a great point. Given that, and transients never being guaranteed to stick around, I'm leaning towards keeping it as simple as possible. 

I like Jan's suggestion there 👍🏻 

---

Another option could be to return to using absolute paths and `[ 'size' => ... ]`, at least for 5.6. A future ticket could work on trimming them, if the tangible impact is worth the added code, risk of bugs, etc.

The suggestion above seems simple enough, though. I'll play around with it first.

Reply:
er, i think the `delete_transient` will need to happen when `$directory_cache[ $directory ]` is detected, like in  bd749dc5e96f6a9e07b1bc50183a45995dba7fbc. now that we're using `WP_CONTENT_DIR`, there shouldn't be any cases where `$cache_path` equals `$directory` <sup>1</sup>. Because of that, `$directory_cache[ $cache_path ]` should never have the `size` array.

I think at this point it might be safer to rely on the upgrade routine to delete the transient, and just have `recurse_dirsize()` ignore any invalid/old data.

`[1]` ...except I don't think we can use `WP_CONTENT_DIR` either, because that's still making assumptions that won't always be true. The `upload_path` option could be an absolute path that's outside of `WP_CONTENT_DIR`, and so could `BLOGUPLOADDIR`. 

I think `_wp_upload_dir()['basedir']` is the most precise and reliable way. We should be able to use that syntax instead of creating a temp variable, since [PHP 5.4 added support for it](https://wiki.php.net/rfc/functionarraydereferencing).

I'm gonna work on a commit to verify all that.

Reply:
Too many edge cases :-) 
Thanks for sorting them all out. And yes, I agree, that WP_CONTENT != upload folders. So many variables containing similar information.

What exactly do you mean with
> i think the delete_transient will need to happen when $directory_cache[ $directory ] is detected, like in bd749dc

Wouldn't that lead to the transient being always deleted? Maybe I just got you wrong, because the commit you referenced runs regardles of the content of the existing cache path and we only want to delete it when a value is found, but it's not an int, right?

Reply:
...except `recurse_dirsize()` can be called for any path, not just an upload (e.g., `WP_DEBUG_DATA::get_sizes()`).

I think it'll be safer to keep the absolute `$directory`, but we can still strip out the `[size]` bit. That actually has the side-benefit of overriding the data too. I've got that mostly working, just wrapping one last unit test.

Will still need some good testing after that, though.

Reply:
er, my last comment was posted before i saw https://github.com/WordPress/wordpress-develop/pull/783#discussion_r535229895, so it was just more stream of consciousness.

4e1c3900601a96193b2be86598cba05cd5d3b26a shows what i'm currently thinking. the transient shouldn't get deleted when it encounters old data

Reply:
i think i've got things working pretty well. the only thing left is that the 5.5 test passes the first time it's ran, but fails on subsequent runs, because the folder isn't being cleaned up like _(i think)_ it should.

Reply:
that's fixed now. i'm gonna resolve this thread since it's getting pretty long, but please open a new one if you see any problems or anything.